src/ddsrt/src/filesystem/windows/filesystem.c

```diff
@@ -0,0 +1,122 @@
+/*
+ * Copyright(c) 2006 to 2018 ADLINK Technology Limited and others
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0, or the Eclipse Distribution License
+ * v. 1.0 which is available at
+ * http://www.eclipse.org/org/documents/edl-v10.php.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR BSD-3-Clause
+ */
+
+#include <string.h>
+
+#include "dds/ddsrt/filesystem.h"
+#include "dds/ddsrt/string.h"
+#include "dds/ddsrt/heap.h"
+#include "dds/ddsrt/io.h"
+
+dds_return_t ddsrt_opendir(const char *name, ddsrt_dirHandle *dir)
+{
+    dds_return_t result;
+
+    TCHAR szDir[DDSRT_PATH_MAX + 1];
+    WIN32_FIND_DATA FileData;
+    HANDLE hList;
+
+    result = DDS_RETCODE_ERROR;
+    if (dir) {
+        snprintf(szDir, DDSRT_PATH_MAX, "%s\\*", name);
+        hList = FindFirstFile(szDir, &FileData);
+
+        if (hList != INVALID_HANDLE_VALUE) {
+            *dir = (ddsrt_dirHandle)hList;
```

Comment:
This cast (and all like it) are also superfluous and had better be removed. (Or, alternatively, one could wrap turn ddsrt_dirHandle into a struct wrapping the underlying object, &c. Your choice.)